subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/metadata/DefaultMavenImmutableAttributesFactory.java

```diff
@@ -101,6 +103,26 @@ public ImmutableAttributes platformWithUsage(ImmutableAttributes original, Strin
         return result;
     }
 
+    @Override
+    public ImmutableAttributes sourcesVariant(ImmutableAttributes original) {
+        ImmutableAttributes result = original;
+        result = concat(result, CATEGORY_ATTRIBUTE, new CoercingStringValueSnapshot(Category.DOCUMENTATION, objectInstantiator));
+        result = concat(result, Bundling.BUNDLING_ATTRIBUTE, objectInstantiator.named(Bundling.class, Bundling.EXTERNAL));
+        result = concat(result, DocsType.DOCS_TYPE_ATTRIBUTE, objectInstantiator.named(DocsType.class, DocsType.SOURCES));
+        result = concat(result, USAGE_ATTRIBUTE, new CoercingStringValueSnapshot("java-runtime", objectInstantiator));
+        return result;
+    }
+
+    @Override
+    public ImmutableAttributes javadocVariant(ImmutableAttributes original) {
+        ImmutableAttributes result = original;
+        result = concat(result, CATEGORY_ATTRIBUTE, new CoercingStringValueSnapshot(Category.DOCUMENTATION, objectInstantiator));
+        result = concat(result, Bundling.BUNDLING_ATTRIBUTE, objectInstantiator.named(Bundling.class, Bundling.EXTERNAL));
+        result = concat(result, DocsType.DOCS_TYPE_ATTRIBUTE, objectInstantiator.named(DocsType.class, DocsType.JAVADOC));
+        result = concat(result, USAGE_ATTRIBUTE, new CoercingStringValueSnapshot("java-runtime", objectInstantiator));
```

Comment:
Can you replace these two places with `Usage.JAVA_RUNTIME`?

Reply:
Done - though I have a small concern with the way we instantiate the "legacy" attributes using constants on the interface and CoercingStringValueSnapshot, vs. `Bundling` and `DocsType` where we use `objectInstantiator` directly. I wonder if the "legacy" values are done that way on purpose for some kind of performance reason?

Reply:
I think this is because objectInstantiator was added later